Index: externals/chease/src-f90/Makefile
===================================================================
--- externals/chease/src-f90/Makefile	(revision 1281)
+++ externals/chease/src-f90/Makefile	(working copy)
@@ -9,6 +9,7 @@
 PROG_TEST_CHEASE = test_chease
 PROG_TEST_CHEASE_ITM = test_chease_itm
 PROG_TEST_CHEASE_IMAS = test_chease_imas
+PROG_MUSCLE   = libchease_muscle
 
 # Define variable CHEASE_F90 which will then be used to defined the compilation flags automatically
 
@@ -47,15 +48,15 @@
 ifeq ($(LIBSlibxml),-lxml2)
 	SRCS_0 = chease_new.f90
 	SRCS_1 = chease_prog_new.f90
-	SRCS_f90_xml_interface = xml_file_reader.f90 xml2eg.f90 assign_chease_codeparameters_choices_xml2eg.f90
-	SRCS_c_xml_interface = xml2eg_c.c
-	INCLUDE_Clibxml = -I/usr/include/libxml2
+	SRCS_f90_xml_interface = assign_chease_codeparameters_choices_xml2eg.f90
+	SRCS_c_xml_interface = 
+	INCLUDE_Clibxml = -I$(INC_XML)
 else
 	SRCS_0 = chease.f90
 	SRCS_1 = chease_prog.f90
 	SRCS_f90_xml_interface = 
 	SRCS_c_xml_interface = 
-	INCLUDE_Clibxml = 
+	INCLUDE_Clibxml = -I$(INC_XML)
 endif
 
 # SRCS_0 chease subroutine in chease.f90
@@ -144,15 +145,17 @@
 SRCS_KEPLER_withlapack	= $(SRCS_0) $(SRCS_all) $(SRCS_0_NOHDF5) $(SRCS_2_ITM) $(SRCS_0_lapackpart)
 # libchease_kepler_imas
 SRCS_KEPLER_IMAS	= $(SRCS_0_IMAS) $(SRCS_all) $(SRCS_0_NOHDF5) $(SRCS_2_IMAS) $(SRCS_0_lapackpart)
+# libchease_muscle
+SRCS_MUSCLE	= $(SRCS_f90_xml_interface) $(SRCS_0) $(SRCS_all) $(SRCS_0_NOHDF5) $(SRCS_2) 
 
 #INCS =	BNDIND.inc COMDAT.inc HERMIT.inc SOLOV.inc mdslib.inc
 INCS =	BNDIND.inc COMDAT.inc HERMIT.inc SOLOV.inc
 
 # make ln -s euitm_schemas_407b.f90 euitm_schemas.f90 before, not needed for euitm_routines.f90
 MODS     = itm_types.f90 euitm_schemas.f90 neobscoeffmod.f90 globals.f90 interpol.f90 prec_const.f90 \
-	   sigmaneomod.f90 string_manipulation_tools.f90 euitm_xml_parser.f90
+	   sigmaneomod.f90 
 MODS_ITM = itm_types.f90 neobscoeffmod.f90 globals.f90 interpol.f90 prec_const.f90 \
-	   sigmaneomod.f90 string_manipulation_tools.f90 euitm_xml_parser.f90
+	   sigmaneomod.f90 
 MODS_IMAS = itm_types.f90 euitm_schemas.f90 neobscoeffmod.f90 globals.f90 interpol.f90 prec_const.f90 \
 	   sigmaneomod.f90 string_manipulation_tools.f90 euitm_xml_parser.f90 itm_ids_utilities.f90
 
@@ -166,6 +169,7 @@
 OBJS_KEPLER = $(MODS_ITM:.f90=.o) $(MODS_interpos:.f90=.o) $(SRCS_KEPLER_f90:.f90=.o) $(SRCS_KEPLER_c:.c=.o)
 OBJS_KEPLER_withlapack = $(MODS_ITM:.f90=.o) $(MODS_interpos:.f90=.o) $(SRCS_KEPLER_withlapack:.f90=.o)
 OBJS_KEPLER_IMAS = $(MODS_IMAS:.f90=.o) $(MODS_interpos:.f90=.o) $(SRCS_KEPLER_IMAS:.f90=.o)
+OBJS_MUSCLE = $(MODS_ITM:.f90=.o) $(MODS_interpos:.f90=.o) $(SRCS_MUSCLE:.f90=.o) $(SRCS_c_xml_interface:.c=.o)
 
 .PRECIOUS:	$(SRCS) $(SRCS_1) $(SRCS_IMAS) $(INCS) $(MODS) $(MODS_ITM) $(MODS_interpos) $(MODS_IMAS) 
 
@@ -235,6 +239,12 @@
 	rm -f $(PROG_KEPLER_IMAS).a
 	@ar -rv $(PROG_KEPLER_IMAS).a $(OBJS_KEPLER_IMAS) $(OBJS_IMASlocal)
 
+$(PROG_MUSCLE): checkCHEASE_F90 checkCHEASE_MACHINE $(OBJS_MUSCLE)
+	@echo "CHEASE_F90: $(CHEASE_F90)"
+	rm -f $(PROG_MUSCLE).a
+	@ar -rv $(PROG_MUSCLE).a $(OBJS_MUSCLE) 
+
+
 $(FUTILS)/lib$(futils).a : 
 	# cp $(FUTILS)/$(Makefile_futils) $(FUTILS)/Makefile
 	@cd $(FUTILS); \
@@ -327,8 +337,8 @@
 apcoef.o: globals.o
 apcoef2.o: globals.o
 #assign_code_parameters.o: prec_const.o euitm_schemas.o euitm_xml_parser.o globals.o
-assign_chease_codeparameters_reflist.o: prec_const.o euitm_xml_parser.o globals.o
-assign_chease_codeparameters_choices_xml2eg.o: prec_const.o globals.o xml2eg.o
+assign_chease_codeparameters_reflist.o: prec_const.o globals.o
+assign_chease_codeparameters_choices_xml2eg.o: prec_const.o globals.o 
 xml2eg.o: xml2eg.f90
 	$(F90) $(F90FLAGS_parser_nor8) -c xml2eg.f90
 atcoef.o: globals.o
@@ -377,7 +387,7 @@
 chease_prog.o: SOLOV.inc COMDAT.inc
 #euitm_schemas.o: 
 #euitm_routines.o: euitm_schemas.o
-euitm_xml_parser.o: itm_types.o string_manipulation_tools.o
+euitm_xml_parser.o: itm_types.o 
 #euitm_utilities.o: euitm_schemas.o
 evlate.o: globals.o
 four1.o: prec_const.o
Index: externals/chease/src-f90/Makefile.define_FLAGS
===================================================================
--- externals/chease/src-f90/Makefile.define_FLAGS	(revision 1281)
+++ externals/chease/src-f90/Makefile.define_FLAGS	(working copy)
@@ -427,3 +427,148 @@
     OBJS_euitm =
   endif
 endif
+
+ifeq ($(CHEASE_MACHINE),linux_nohdf5)
+  ifeq ($(CHEASE_F90),gfortran)
+    F90 = gfortran
+    F90FLAGS = -g -O2 -cpp -fPIC -ffree-line-length-none  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -g -O2 -fPIC -ffree-line-length-none  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser_nor8 = -g -O2 -fPIC -ffree-line-length-none -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos)
+    LDFLAGS = -g -O2 -fPIC -ffree-line-length-none -I$(INC_UAL) 
+    FUTILS = $(PWD)/externals/chease/src-f90/futils_src
+    INCL_FUTILS= -I$(FUTILS)
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L/usr/lib/x86_64-linux-gnu/lapack -llapack -lrefblas
+    LIBSlibxml = -lxml2
+    OBJS_euitm =
+  endif
+endif
+
+ifeq ($(CHEASE_MACHINE),GW)
+  ifeq ($(CHEASE_F90),ifort)
+    F90 = ifort
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser_nor8 = -O1 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos)
+    LDFLAGS = -O2 -g -r8 -fPIC -I$(INC_UAL) 
+    FUTILS = $(PWD)/futils_src
+    INCL_FUTILS= -I$(FUTILS)
+    HDF5=
+    INCL_HDF5 = -I$(HDF5)/include
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L/afs/ipp/itm/switm/lib/lapack/3.4.2/lib/amd64_intel_12 -llapack -lrefblas
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS)
+    LIBSlibxml = -lxml2
+    OBJS_euitm =
+  endif
+endif
+
+ifeq ($(CHEASE_MACHINE),helios) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_DIR)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_DIR)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+
+ifeq ($(CHEASE_MACHINE),marconi) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_HOME)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_HOME)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+
+ifeq ($(CHEASE_MACHINE),supermuc) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_BASE)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_BASE)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+
+ifeq ($(CHEASE_MACHINE),psnc) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(hdf5_root)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(hdf5_root)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+
+ifeq ($(CHEASE_MACHINE),hydra) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_HOME)/include 
+    LIBS =  -L$(LIB_UAL) -lual -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_HOME)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+
+ifeq ($(CHEASE_MACHINE),draco) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_HOME)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_HOME)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+ifeq ($(CHEASE_MACHINE),cobra) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_HOME)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_HOME)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
+
+ifeq ($(CHEASE_MACHINE),neale) 
+  ifeq ($(CHEASE_F90),ifort) 
+    F90 = ifort 
+    F90FLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    F90FLAGS_parser = -O1 -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML) -I$(DIR_interpos) 
+    LDFLAGS = -O2 -g -r8 -fPIC  -I$(INC_UAL) -I$(INC_XML)
+    FUTILS = $(PWD)/futils_src 
+    INCL_FUTILS= -I$(FUTILS) 
+    INCL_HDF5 = -I$(HDF5_HOME)/include 
+    LIBS =  -L$(LIB_UAL) -lual -L$(LIB_XML) -lxmllib -lz -L$(MKL_LIB) -llapack -lblas 
+    LIBS_HDF5 =  -L$(FUTILS) -lfutils -L$(HDF5_HOME)/lib -lhdf5_fortran -lhdf5 -lz $(LIBS) 
+    LIBSlibxml = -lxml2
+    OBJS_euitm = 
+  endif 
+endif 
Index: externals/chease/src-f90/Makefile.define_MACHINE
===================================================================
--- externals/chease/src-f90/Makefile.define_MACHINE	(revision 1281)
+++ externals/chease/src-f90/Makefile.define_MACHINE	(working copy)
@@ -60,4 +60,34 @@
  ifneq (,$(findstring yes,$(ITM_ENVIRONMENT_LOADED)))
     # overwrites the CHEASE_MACHINE set in Makefile.define_MACHINE by using hostname
     CHEASE_MACHINE = gateway_with_itm
  endif
+ ifeq ($(SYS),GW)
+   CHEASE_MACHINE = GW
+ endif
+ ifeq ($(SYS),HELIOS)
+   CHEASE_MACHINE = helios
+ endif
+ ifeq ($(SYS),MARCONI)
+   CHEASE_MACHINE = marconi
+ endif
+ ifeq ($(SYS),SUPERMUC)
+   CHEASE_MACHINE = supermuc
+ endif
+ ifeq ($(SYS),INULA)
+   CHEASE_MACHINE = psnc
+ endif
+ ifeq ($(SYS),EAGLE)
+   CHEASE_MACHINE = psnc
+ endif
+ ifeq ($(SYS),HYDRA)
+   CHEASE_MACHINE = hydra
+ endif
+ ifeq ($(SYS),DRACO)
+   CHEASE_MACHINE = draco
+ endif
+ ifeq ($(SYS),COBRA)
+   CHEASE_MACHINE = cobra
+ endif
+ ifeq ($(SYS),NEALE)
+   CHEASE_MACHINE = neale
+ endif
+ ifeq ($(SYS),Default)
+   CHEASE_MACHINE = linux_nohdf5
+ endif
