Index: externals/gem/actor/gem.F90
===================================================================
--- externals/gem/actor/gem.F90	(revision 107)
+++ externals/gem/actor/gem.F90	(working copy)
@@ -371,6 +371,11 @@
   END IF
 #endif
 
+
+  call deallocate_cpo(coreprof)
+  call deallocate_cpo(eq)
+
+
 210 FORMAT('Ftube ',i2,' step ',3i4,' tau = ',g12.4)
 310 FORMAT('run exploded at time = ',g11.5,' step ',3i4,'  taumin = ',g9.3)
 320 FORMAT('Ftube ',i2,' cascade at time = ',g11.5)
Index: externals/gem/actor/itm_cpos.h90
===================================================================
--- externals/gem/actor/itm_cpos.h90	(revision 107)
+++ externals/gem/actor/itm_cpos.h90	(working copy)
@@ -6,6 +6,7 @@
   USE Read_structures
   USE Write_structures
   USE Copy_structures
+  USE Deallocate_structures
 
   IMPLICIT NONE
 
